### PR TITLE
Feature - Support for subdomains on origin

### PR DESCRIPTION
**Pull request by vigan-abd**
### Description:
This PR extends the library's ability to use also different `origin` from `rpId`, e.g. using `https://www.example.com` as origin when rpId is `example.com`. It simply adds an optional `rpId` field in expected result classes which would be used to verify `rpIdHash` instead of origin when present. Since `rpId` field is optional if omitted verify `rpIdHash` would use `origin` instead so this way is fully backward compatible with older versions of lib.

Example usage on register challenge verification:
```js
const expected = {
  challenge: "LN6BmfUMFzBgfmqoaAU26k2Ef7STFrZ5eHF8P_q309usd_Vj-qMurZaKLVvW60OE31l_3jKucjQ_1kA0mp5aNA",
  origin: "https://app.test.com", // <- this one is verified against client data origin
  rpId: 'test.com', // <- this one is used for verifying rpIdHash instead of client data origin
  factor: "either"
}

const regRes = await f2l.attestationResult(regResponse, expected)
```

### Breaking changes:
- [ ]

### New features:
- support for subdomains on origin

### Fixes:
- TS doc fixes
- removed package-lock.json which caused travis failures

### PR status:
- [x] Version bumped
- [x] Change-log updated
- [x] Tests added or updated

### CODE REVIEW

**Comment by vigan-abd** [comment]
Hi @JamesCullum, thanks again for maintaining this really useful lib, could you please check this PR and deploy a new version if it's ok?

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,5 @@ npm-debug.log
 coverage
 *~
 .idea
-.nyc_output
\ No newline at end of file
+.nyc_output
+package-lock.json
```
Comment (JamesCullum) [inline .gitignore]: package-lock.json needs to be committed to allow fixed versions, eg for a reconstructable CI run

File: lib/validator.js [hunk 2]
```diff
@@ -39,17 +40,20 @@ async function validateExpectations() {
 		}
 	}
 
-	if (req.size !== exp.size) {
```
Comment (JamesCullum) [inline lib/validator.js:42]: Please don't remove code, but make it work with your use case properly

File: test/utilsTest.js [hunk 3]
```diff
@@ -34,13 +35,13 @@ describe("utils", function() {
 		it("throws on undefined origin", function() {
 			assert.throws(() => {
 				checkOrigin(undefined);
-			}, Error, "Invalid URL: undefined");
+			}, Error, "Invalid URL");
```
Comment (JamesCullum) [inline test/utilsTest.js:38]: Why remove the specification of which URL it was? This would be useful for debugging, isnt it?

File: package.json [hunk 4]
```diff
@@ -55,7 +55,8 @@
 		"node-jose": "^2.0.0",
 		"node-webcrypto-ossl": "^1.0.49",
 		"pkijs": "=2.1.58",
-		"psl": "^1.1.24"
+		"psl": "^1.1.24",
```
Comment (JamesCullum) [inline package.json:58]: For which functions are those added dependencies needed?

File: test/validatorTest.js [hunk 5]
```diff
@@ -66,16 +66,6 @@ describe("attestation validation", function() {
 			return assert.isRejected(attResp.validateExpectations(), Error, "expectations should be of type Map");
 		});
 
-		it("throws if too many expectations", function() {
```
Comment (JamesCullum) [inline test/validatorTest.js:69]: As above, shouldnt be removed

File: types/index.d.ts [hunk 6]
```diff
@@ -63,8 +63,8 @@ declare module "fido2-library" {
   }
 
   interface ExpectedAttestationResult {
-    rpId: string;
-    origin?: string | string[];
+    rpId?: string;
+    origin: string;
```
Comment (JamesCullum) [inline types/index.d.ts:67]: This looks like a breaking change to me

**Review by JamesCullum — CHANGES_REQUESTED** [review]
Please integrate this feedback

**Comment by JamesCullum** [comment]
Thanks for your contribution. Please make sure that test coverage only increases and doesn't decrease

File: types/index.d.ts (hunk 6 above)
Comment (vigan-abd) [reply types/index.d.ts:67]: well this one was just adjustment according to js code, e.g. if you tried to provide origin as string array it would throw an error on validation, that's why I adjusted it to be just a string. Do you think that I should restore it?

File: test/utilsTest.js (hunk 3 above)
Comment (vigan-abd) [reply test/utilsTest.js:38]: this was causing build failures in node v14 since there native URL class doesn't include invalid url on error message, I've adjusted it to be compatible with node v14 and above, should I restore it?

tldr:
error message thrown on node v14: `Invalid URL`
error message thrown on node v15: `Invalid URL: asdf`

File: lib/validator.js (hunk 2 above)
Comment (vigan-abd) [reply lib/validator.js:42]: returned old code and tests, now it just excludes optional fields from size check

File: test/validatorTest.js (hunk 5 above)
Comment (vigan-abd) [reply test/validatorTest.js:69]: restored :)

File: .gitignore (hunk 1 above)
Comment (vigan-abd) [reply .gitignore]: ok managed to produce builds with package-lock.json also by adding some missing dev dependencies, now package-lock.json is restored :+1:

**Comment by vigan-abd** [comment]
hi @JamesCullum , thanks for the review, submitted feedback changes and answered some questions, could you check it again please?

File: types/index.d.ts (hunk 6 above)
Comment (JamesCullum) [reply types/index.d.ts:67]: I was mainly talking about the question mark - if I remember correctly this means that before origin was optional to provide, while now it isnt. This might break some scripts, wouldn't it?

File: test/utilsTest.js (hunk 3 above)
Comment (JamesCullum) [reply test/utilsTest.js:38]: Didn't know that this was the behavior in node v14 - just to be clear, we are not talking about the versions above, right? But it doesn't matter that much, as I was mostly worrying about debugging information, but missed that it still has its own test case.

File: types/index.d.ts (hunk 6 above)
Comment (vigan-abd) [reply types/index.d.ts:67]: Actually origin isn't optional in latest package, it's required and type must be string. If you try to omit it in latest version (master branch) of package or provide it as string array then it would throw an error.
On the other side in latest version if you try to provide rpId value it would throw an error, so index.d.ts is actually outdated based on js codebase in master branch

File: lib/validator.js [hunk 7]
```diff
@@ -39,7 +40,9 @@ async function validateExpectations() {
...
+	let expSize = exp.size;
+	if (exp.has("rpId")) expSize--; // don't include optional fields
```
Comment (JamesCullum) [inline lib/validator.js]: This feels like we should find a more beautiful solution, eg making a list of optional fields (referencing it in the documentation) and count all non-optional fields, what do you think? This way we can more easily extend this later.

Comment (vigan-abd) [reply lib/validator.js]: agree with you, it would be more clean and maintainable, doing it now

**Review by JamesCullum — COMMENTED** [review]
Thanks a lot for your quick changes, it looks a lot better now! Lets solve the two threads regarding the TypeScript and the nicer version of counting optional fields, then I think we're good to go

**Comment by vigan-abd** [comment]
thanks for the review again, added optional fields as suggested, now we have to solve only this one I guess https://github.com/FIDO-Tools/fido2-library/pull/14#discussion_r640112501

File: types/index.d.ts (hunk 6 above)
Comment (vigan-abd) [reply types/index.d.ts:67]: examples from master branch:

```
  const expectedReg = {
    challenge: "LN6BmfUMFzBgfmqoaAU26k2Ef7STFrZ5eHF8P_q309usd_Vj-qMurZaKLVvW60OE31l_3jKucjQ_1kA0mp5aNA",
    origin: ["https://example.com"],
    factor: "either"
  }
  regResponse.id = Buffer.from(regResponse.id, 'base64').buffer

  const regRes = await f2l.attestationResult(regResponse, expectedReg)
  console.log(true)

# throws:
TypeError: expected 'origin' should be string, got object
    at Object.parseExpectations (...../playground/node_modules/fido2-library/lib/parser.js:26:10)
```

```
  const expectedReg = {
    challenge: "LN6BmfUMFzBgfmqoaAU26k2Ef7STFrZ5eHF8P_q309usd_Vj-qMurZaKLVvW60OE31l_3jKucjQ_1kA0mp5aNA",
    rpId: "example.com",
    factor: "either"
  }
  regResponse.id = Buffer.from(regResponse.id, 'base64').buffer

  const regRes = await f2l.attestationResult(regResponse, expectedReg)
  console.log(true)

# throws:
Error: expectation did not contain value for 'origin'
    at Fido2AttestationResult.validateExpectations (..../playground/node_modules/fido2-library/lib/validator.js:38:10)
```

Comment (vigan-abd) [reply types/index.d.ts:67]: As you can see from examples above if we make requests according to original index.d.ts we would get errors since it doesn't comply with actual js code:
```ts
  interface ExpectedAttestationResult {
    rpId: string;
    origin?: string | string[];
    challenge: string;
    factor: Factor;
  }
```

File: lib/validator.js [hunk 8]
```diff
@@ -34,16 +35,29 @@ async function validateExpectations() {
...
-	if (req.size !== expSize) {
-		throw new Error(`wrong number of expectations: should have ${req.size} but got ${exp.size}`);
+	let optCount = 0;
+	for (const [field] of exp) {
+		if (opt.has(field)) {
+			optCount++;
+		}
+	}
+
+	if (req.size !== exp.size - optCount) {
```
Comment (JamesCullum) [inline lib/validator.js:59]: Wouldnt it make more sense to count the mandatory fields and compare it without the math?

File: types/index.d.ts (hunk 6 above)
Comment (JamesCullum) [reply types/index.d.ts:67]: I see - thanks for the good explanation!

File: lib/validator.js (hunk 8 above)
Comment (vigan-abd) [reply lib/validator.js:59]: well we compare the mandatory fields before doing this math:
```js
	for (let field of req) {
		if (!exp.has(field)) {
			throw new Error(`expectation did not contain value for '${field}'`);
		}
	}
```

The intention of this check here is to assure that no unknown field is included. In main branch since it doesn't have optional fields it simply compares if exp has more values than required, however in our case since we added optional fields we have to check it against all exp fields - optional found fields.

E.g. this change would catch the following cases:
```js
{
    challenge: "LN6BmfUMFzBgfmqoaAU26k2Ef7STFrZ5eHF8P_q309usd_Vj-qMurZaKLVvW60OE31l_3jKucjQ_1kA0mp5aNA",
    rpId: 'exampl.com',
    origin: 'https://www.example.com',
    factor: "either",
    foo: "bar" // <- invalid
  }
```

```js
{
    challenge: "LN6BmfUMFzBgfmqoaAU26k2Ef7STFrZ5eHF8P_q309usd_Vj-qMurZaKLVvW60OE31l_3jKucjQ_1kA0mp5aNA",
    origin: 'https://www.example.com',
    factor: "either",
    foo: "bar" // <- invalid
  }
```

Comment (JamesCullum) [reply lib/validator.js:59]: Sure, what I meant was that I think it would make more sense to count the found mandatory fields in the loop above and then check if this number matches - this is easier to understand and has less potential side effects

Comment (vigan-abd) [reply lib/validator.js:59]: yep but in that case we wouldn't need to compare counts since it wouldn't reach here at all if at least one required field isn't found due to thrown error (if (!exp.has(field)) throw new Error(`expectation did not contain value for '${field}'`)).

also counting required fields wouldn't check for other fields that were provided but are unknown (neither optional or required)

Comment (vigan-abd) [reply lib/validator.js:59]: Hi @JamesCullum, could you please check the comment above?

**Review by JamesCullum — APPROVED** [review]
